app/src/main/java/com/github/scribeWizTeam/scribewiz/transcription/PitchDetector.kt

```diff
@@ -0,0 +1,87 @@
+package com.github.scribeWizTeam.scribewiz.transcription
+
+import kotlin.math.*
+
+// for pitch detection:
+// https://en.wikipedia.org/wiki/Pitch_detection_algorithm
+
+typealias Signal = List<Double>
+typealias Frequency = Double
+typealias Energy = Double
+
+class PitchDetector(val samplingFreq: Frequency){
+
+    init {
+        if (samplingFreq <= 0){
+            throw IllegalArgumentException(
+                "sampling frequency of PitchDetector should be positive"
+            )
+        }
+    }
+
+    private fun autocorrelation(signal: Signal, lag: Int): Energy{
+        var energy = 0.0
+        for (i in 0 until signal.size - lag){
+            energy += signal[i]*signal[i+lag]
+        }
+        return energy
+    }
+
+    private fun square_sum(signal: Signal, lag: Int): Energy{
+        var energy = 0.0
+        for (i in 0 until signal.size - lag){
+            energy += signal[i].pow(2)+signal[i+lag].pow(2)
```

Comment:
why do we sum up signal at [i] and signal at [i + lag] ?

Reply:
That’s one of the steps for the signal processing algorithm I used.
The algorithm is described in details in this paper: http://www.cs.otago.ac.nz/tartini/papers/A_Smarter_Way_to_Find_Pitch.pdf

The idea is that it comes from the expansion of (x(t) - x(t+τ))^2 which gives you x(t)^2 + x(t+τ)^2 - 2x(t)x(t+τ)
The second part is 2 times the signal autocorrelation and what remains is this `square_sum` function.